pywbem_mock/_defaultinstanceprovider.py

```diff
@@ -0,0 +1,439 @@
+
+"""
+
+from __future__ import absolute_import, print_function
+
+import six
+
+from pywbem import CIMInstance, CIMInstanceName, CIMError, \
+    CIM_ERR_NOT_FOUND, CIM_ERR_INVALID_PARAMETER, CIM_ERR_INVALID_CLASS, \
+    CIM_ERR_ALREADY_EXISTS
+
+from pywbem._utils import _format
+
+from ._baseprovider import BaseProvider
+
+# None of the request method names conform since they are camel case
+# pylint: disable=invalid-name
+
+__all__ = ['DefaultInstanceProvider']
+
+
+def validate_inst_props(namespace, target_class, instance):
+    """
+    Validate that all properties in instance are in target_class and validate
+    that property types are compatible.
+    """
+    # TODO: This method modifies instance.
+    for iprop_name in instance:
+        if iprop_name not in target_class.properties:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Property {0!A} specified in NewInstance is not "
+                        "exposed by class {1!A} in namespace {2!A}",
+                        iprop_name, target_class.classname, namespace))
+
+        cprop = target_class.properties[iprop_name]
+        iprop = instance.properties[iprop_name]
+        if iprop.is_array != cprop.is_array or \
+                iprop.type != cprop.type:
+            raise CIMError(
+                CIM_ERR_INVALID_PARAMETER,
+                _format("Instance and class property {0!A} types "
+                        "do not match: instance={1!A}, class={2!A}",
+                        iprop_name, iprop, cprop))
+
+        # The class and instnames are the same except for possible case
+        # sensitivity. If case different, set cprop_name into new instance
+        # to maintain case equality
+        cprop_name = cprop.name
+        if cprop_name != iprop_name:
+            instance.properties[iprop_name].name = cprop_name
+
+        # If property not in instance, add it from class and use default value
+        # from class
+        for cprop_name in target_class.properties:
+            if cprop_name not in instance:
+                default_value = target_class.properties[cprop_name]
+                instance[cprop_name] = default_value
+
+
+class ProviderDispatcher(BaseProvider):
+    """
+    This class dispatches requests destined for the instance provider methods
+    defined in DefaultInstanceProvider  to the either the default instance
+    provder method or the provider defined for processing the defined class in
+    the defined namespace.
+    """
+    def __init__(self, conn, default_instance_provider):
+        """
+        """
+        # TODO: This not correct way to make cimrepository available to
+        # multiple subclasses
+        super(ProviderDispatcher, self).__init__(
+            conn=None, cimrepository=None)
+        self.default_instance_provider = default_instance_provider
+
+        self.cimrepository = conn.cimrepository
+        self.provider_registry = conn.provider_registry
```

Comment:
Instead of passing in `conn` and then using exactly two attributes from it, I suggest to pass in these two attributes. That avoids the back reference and makes it more clear what information is actually passed.

Reply:
This is getting messy.  First, we have to pass cim_repository to each of the subclasses but some form of self.cim_repository must exist in BaseProvider just to avoid lint errors since it is referenced.

I am passing conn to the main provider only so it can be passed to the BaseProvider just to extract host and disable_pull_operations which are used in some of the methods.

Proposed solution.

1. MainProvider pass host, disable_pull_operation, cim_repository to MainProvider. 

2. DefaultInstanceProvider Pass cim_repository

3. ProviderDispatcher .CIMRepository, provider_registry, instance of DefaultInstanceProvider

**DONE:** For next commit